pkg/utils/gateway_wrapper.go

```diff
@@ -0,0 +1,80 @@
+package utils
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	MultiClusterIPAddressType       gatewayv1beta1.AddressType = "kuadrant.io/MultiClusterIPAddress"
+	MultiClusterHostnameAddressType gatewayv1beta1.AddressType = "kuadrant.io/MultiClusterHostnameAddress"
+)
+
+type GatewayWrapper struct {
+	*gatewayv1beta1.Gateway
+	isMultiCluster bool
+}
+
+func NewGatewayWrapper(g *gatewayv1beta1.Gateway) (*GatewayWrapper, error) {
+	gw := &GatewayWrapper{Gateway: g, isMultiCluster: false}
+
+	for i, address := range gw.Status.Addresses {
+		if i == 0 {
+			gw.isMultiCluster = isMultiClusterAddressType(*address.Type)
+			continue
+		}
+		if gw.isMultiCluster == isMultiClusterAddressType(*address.Type) {
+			continue
+		}
+		return nil, fmt.Errorf("gateway is invalid: inconsistent status addresses")
+
+	}
+	return gw, nil
+}
+
+func isMultiClusterAddressType(addressType gatewayv1beta1.AddressType) bool {
+	return addressType == MultiClusterIPAddressType || addressType == MultiClusterHostnameAddressType
+}
+
+func (a *GatewayWrapper) GetKind() string {
+	return "GatewayWrapper"
+}
+
+func (a *GatewayWrapper) IsMultiCluster() bool {
+	return a.isMultiCluster
+}
```

Comment:
I think we would be better to split the validation part from the check for multi cluster. Reason being we need to do other validation of the gateway, and it would all be better done at the start as part of an explicit action. I'm thinking something like:

```
func (g *GatewayWrapper) Validate() error {
	for _, address := range g.Status.Addresses {
		if isMultiClusterAddress(address) != g.IsMultiCluster() {
			return fmt.Errorf("gateway is invalid: inconsistent status addresses")
		}
	}
	return nil
}

func (g *GatewayWrapper) IsMultiCluster() bool {
	if len(g.Status.Addresses) > 0 {
		return isMultiClusterAddress(g.Status.Addresses[0])
	}
	return false
}
```

Reply:
done